File: src/java/src/main/java/dab/Bot.java
package dab;

import java.util.Comparator;
import java.util.List;

public final class Bot {

    private Bot() {}

    public static BotMove greedySafe(State s) {
        if (Core.gameOver(s)) throw new IllegalStateException("game over");
        List<Edge> moves = Core.possibleMoves(s);
        if (moves.isEmpty()) throw new IllegalStateException("no moves");
        int bestClosed = -1;
        int bestRisk = Integer.MAX_VALUE;
        Edge bestEdge = moves.get(0);
        for (Edge e : moves) {
            var res = Core.applyMove(s, e);
            int closed = res.closedBoxes().size();
            int risk = thirdSideRisk(s, e);
            if (closed > bestClosed
                || (closed == bestClosed && risk < bestRisk)
                || (closed == bestClosed && risk == bestRisk && compareEdge(e, bestEdge) < 0)) {
                bestClosed = closed;
                bestRisk = risk;
                bestEdge = e;
            }
        }
        String reason = bestClosed >= 2 ? "greedy_close_2"
            : bestClosed == 1 ? "greedy_close_1"
            : bestRisk == 0 ? "safe_no_third_side"
            : "fallback_min_risk";
        return new BotMove(bestEdge, reason);
    }

    public static int thirdSideRisk(State s, Edge e) {
        var with = new java.util.HashSet<>(s.edges());
        with.add(Edge.normalized(e.a(), e.b()));
        int count = 0;
        for (int[] box : Core.adjacentBoxesForEdge(s.nx(), s.ny(), e)) {
            var be = Core.boxEdges(box[0], box[1]);
            long three = be.stream().filter(with::contains).count();
            if (three == 3) count++;
        }
        return count;
    }

    private static int compareEdge(Edge a, Edge b) {
        int c = Integer.compare(a.a().x(), b.a().x());
        if (c != 0) return c;
        c = Integer.compare(a.a().y(), b.a().y());
        if (c != 0) return c;
        c = Integer.compare(a.b().x(), b.b().x());
        if (c != 0) return c;
        return Integer.compare(a.b().y(), b.b().y());
    }

    public record BotMove(Edge edge, String reason) {}
}

File: src/python/dab/bot.py
# dab/bot.py — greedy-safe move selection
from typing import List, Tuple

from dab.types import Edge, Reason
from dab.state import State
from dab.core import apply_move, possible_moves, game_over
from dab.geometry import adjacent_boxes_for_edge, box_edges


def third_side_risk(state: State, edge: Edge) -> int:
    """Number of adjacent boxes that would have exactly 3 sides after playing edge."""
    edges_with = state.edges | {edge}
    count = 0
    for box in adjacent_boxes_for_edge(state.nx, state.ny, edge):
        be = box_edges(box[0], box[1])
        if sum(1 for e in be if e in edges_with) == 3:
            count += 1
    return count


def greedy_safe_bot_move(state: State) -> Tuple[Edge, Reason]:
    """Best move by policy: greedy_close > safe_no_third_side > fallback_min_risk. Tie: min edge."""
    if game_over(state):
        raise ValueError("game over")
    moves = possible_moves(state)
    if not moves:
        raise ValueError("no moves")
    candidates: List[Tuple[int, int, Edge]] = []
    for e in moves:
        _, closed, _ = apply_move(state, e)
        closed_count = len(closed)
        risk = third_side_risk(state, e)
        candidates.append((-closed_count, risk, e))
    candidates.sort(key=lambda t: (t[0], t[1], t[2]))
    neg_closed, risk, best_move = candidates[0]
    if neg_closed <= -2:
        best_reason: Reason = "greedy_close_2"
    elif neg_closed == -1:
        best_reason = "greedy_close_1"
    elif risk == 0:
        best_reason = "safe_no_third_side"
    else:
        best_reason = "fallback_min_risk"
    return best_move, best_reason
